### PR TITLE
Images

### PR DESCRIPTION
**Pull request by JouvinLea**
Try to make the class works again...
For the moment I am using the script test_curve.py in gammapy/example to test. I will do a test when I would have fix the bug

### CODE REVIEW

**Comment by cdeil** [comment]
Thanks! Let me know if you have any questions or want me to look at something.

File: gammapy/image/maps.py [hunk 1]
```diff
@@ -260,8 +260,8 @@ def solid_angle(self):
         """
         Solid angle image
         """
-        xsky, ysky = self.coordinates(mode='edges')
```
Comment (JouvinLea) [inline gammapy/image/maps.py]: @cdeil 
Get an error:
```
ValueError: too many values to unpack
```
with the old version. You changed the coordinates method of the skymap that return a tuple if coord_type=galactic but an array if coord_type=skycoord.
Is it ok what I did?

Comment (cdeil) [inline gammapy/image/maps.py]: @adonath - Can you have a look and answer this question?

File: gammapy/image/maps.py [hunk 2]
```diff
@@ -260,8 +260,8 @@ def solid_angle(self):
...
-        omega = -np.diff(xsky, axis=1)[1:, :] * np.diff(ysky, axis=0)[:, 1:]
+        coord = self.coordinates(mode='edges')
+        omega = -np.diff(coord.l, axis=1)[1:, :] * np.diff(coord.b, axis=0)[:, 1:]
```
Comment (adonath) [inline gammapy/image/maps.py]: I think this will fail if the `wcs` is not in specified in galactic coordinates. It would be better to use `coordinates(coord_type='galactic', mode='edges')` in `solid_angle()` which behaves the "old way". Could you add a minimal test to `gammapy/image/tests/test_maps.py`? Thanks!

Comment (cdeil) [inline gammapy/image/maps.py]: I think the way to access lon / lat on a SkyCoord in any frame is via `coord.data.lon` and `coord.data.lat`. Would that work here?

```
In [2]: from astropy.coordinates import SkyCoord

In [3]: SkyCoord(42, 43, unit='deg')
Out[3]: 
<SkyCoord (ICRS): (ra, dec) in deg
    (42.0, 43.0)>

In [4]: s = SkyCoord(42, 43, unit='deg')

In [5]: s.l
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-5-bb9361e12b88> in <module>()
----> 1 s.l

/Users/deil/software/anaconda/lib/python3.4/site-packages/astropy/coordinates/sky_coordinate.py in __getattr__(self, attr)
    434         # Fail
    435         raise AttributeError("'{0}' object has no attribute '{1}'"
--> 436                              .format(self.__class__.__name__, attr))
    437 
    438     def __setattr__(self, attr, val):

AttributeError: 'SkyCoord' object has no attribute 'l'

In [6]: s.ra
Out[6]: <Longitude 42.0 deg>

In [7]: s.data
Out[7]: 
<UnitSphericalRepresentation (lon, lat) in deg
    (42.0, 43.0)>

In [8]: s.data.lon
Out[8]: <Longitude 42.0 deg>

In [9]: s.galactic.data.lon
Out[9]: <Longitude 2.5230336177186996 rad>
```

File: gammapy/image/tests/test_maps.py [hunk 3]
```diff
@@ -45,6 +45,10 @@ def test_coordinates(self):
         assert_allclose(coordinates[0][100, 100].degree, 0.01)
         assert_allclose(coordinates[1][100, 100].degree, 0.01)
 
+    def test_solid_angle(self):
```
Comment (JouvinLea) [inline gammapy/image/tests/test_maps.py]: @adonath 
add this test for the solid angle, why is it negative?

File: gammapy/scripts/tests/test_image_pipe.py [hunk 4]
```diff
@@ -0,0 +1,76 @@
+"""Example how to make an acceptance curve and background model image.
```
Comment (JouvinLea) [inline gammapy/scripts/tests/test_image_pipe.py]: @cdeil 
add this test for the map! It doesn't really test somethings on the final values but at least next time if you change some variable names in gammapy we will see it!

File: gammapy/image/tests/test_maps.py (hunk 3 above)
Comment (adonath) [inline gammapy/image/tests/test_maps.py]: That's strange...it shouldn't be negative and it isn't the case for the test files I used. But let's change to make it return safe results...

File: gammapy/image/maps.py [hunk 5]
```diff
@@ -260,7 +260,7 @@ def solid_angle(self):
         """
         Solid angle image
         """
-        xsky, ysky = self.coordinates(mode='edges')
+        xsky, ysky = self.coordinates(coord_type='galactic', mode='edges')
         omega = -np.diff(xsky, axis=1)[1:, :] * np.diff(ysky, axis=0)[:, 1:]
```
Comment (adonath) [inline gammapy/image/maps.py]: Can you add a `np.abs()` call here and remove the minus sign?

File: gammapy/image/maps.py [hunk 6]
```diff
@@ -260,7 +260,7 @@ def solid_angle(self):
...
         return Quantity(omega, 'deg2').to('sr')
```
Comment (adonath) [inline gammapy/image/maps.py]: `omega` should already be a `Quantity` now, could remove the `Quantity` statement and just return `omega` or `omega.to('sr')` here?

File: gammapy/image/maps.py (hunk 5 above)
Comment (cdeil) [inline gammapy/image/maps.py]: Just to confirm what @adonath  is suggesting: just call `np.abs` to make it work for all images. Some images will have positive or negative values for CDELT and the easiest way to handle all images is to take the abs value.

**Comment by cdeil** [comment]
(I'm unsubscribing from this PR now, leaving the review to @adonath. If there's something where you want my input, please ping me directly.)

File: gammapy/scripts/image_pipe.py [hunk 7]
```diff
@@ -285,7 +285,7 @@ def make_exposure(self, obs_id, spectral_index=2.3, for_integral_flux=False):
         xpix_coord_grid, ypix_coord_grid = exposure.coordinates(coord_type='pix')
         # calculate pixel offset from center (in world coordinates)
         coord = pixel_to_skycoord(xpix_coord_grid, ypix_coord_grid, exposure.wcs, origin=0)
-        center = SkyCoord.from_pixel(self.header["NAXIS1"] / 2., self.header["NAXIS2"] / 2., exposure.wcs)
```
Comment (adonath) [inline gammapy/scripts/image_pipe.py]: Note, that the definition of `SkyMap.center()` might be slightly different (see [here](https://github.com/gammapy/gammapy/blob/master/gammapy/image/maps.py#L271)) from the one used here. Is this covered by a test somehow? If not, could you add at least one assert to the test below? E.g. check that the center coordinates are right...because it might affect science results.

**Comment by adonath** [comment]
@leajouvin Thanks for the update of `ImageAnalysis` to the changes I've made to `SkyMap`. I've left a few minor comments. The only major point I got is, we should make sure, that the definition of the center of a `SkyMap` is technically correct. But I'll open a separate issue for this.  

Right now `test_image_pipe.py` fails on Travis-CI. Does it pass locally on your machine?

File: gammapy/image/tests/test_maps.py [hunk 8]
```diff
@@ -103,3 +107,24 @@ def test_reproject(self):
...
+    def Crab_coord(self):
+        coord = SkyCoord(83.63, 22.01, unit='deg').galactic
+        return coord
+
+    def setup(self):
+        center = self.Crab_coord()
+        self.skymap  = SkyMap.empty(nxpix=250, nypix=250, binsz=0.02, xref=center.l.deg,
+                         yref=center.b.deg, proj='TAN', coordsys='GAL')
+
+    def test_center(self):
```
Comment (JouvinLea) [inline gammapy/image/tests/test_maps.py]: @adonath 
add this test for the center by creating an image centered on the Crab position:
```
In [17]: SkyCoord.from_name("Crab").galactic
Out[17]: 
<SkyCoord (Galactic): (l, b) in deg
    (184.55745788, -5.78435671)>
```
When you use the method center of the skymap it gives:
```
<SkyCoord (Galactic): (l, b) in deg
    (184.55974014, -5.78918015)>
```
when you use the line I wrote before:
center = SkyCoord.from_pixel(self.header["NAXIS1"] / 2., self.header["NAXIS2"] / 2., exposure.wcs)
it gives:
```
<SkyCoord (Galactic): (l, b) in deg
    (184.54968906, -5.77918006)>
```

**Comment by JouvinLea** [comment]
@adonath 
I took your suggestions into account!!
Travis failed also locally but not for the test I added... Since I didn't code the other tests I don't know why they failed and I didn't want to touch them just to have the good value...

File: gammapy/scripts/tests/test_image_pipe.py [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+from gammapy.image import bin_events_in_image, disk_correlate, SkyMap, ExclusionMask
+from gammapy.background import fill_acceptance_image
+from gammapy.region import SkyCircleRegion
+from gammapy.stats import significance
+from gammapy.background import OffDataBackgroundMaker
+from gammapy.scripts import ImageAnalysis
+
+
+
+
```
Comment (adonath) [inline gammapy/scripts/tests/test_image_pipe.py]: But this test requires `gammapy-extra`, right? So just add `@requires_data('gammapy-extra')`. This might fix the Travis-CI build as well.

File: gammapy/image/tests/test_maps.py [hunk 10]
```diff
@@ -103,3 +105,25 @@ def test_reproject(self):
         skymap_1.fill(1)
         skymap_1_repr = skymap_1.reproject(skymap_2)
         assert_allclose(skymap_1_repr.data, np.full((100, 100), 1))
+
+
+@requires_data('gammapy-extra')
```
Comment (adonath) [inline gammapy/image/tests/test_maps.py]: This test doesn't require `gammpy-extra`...just remove.

File: gammapy/scripts/image_pipe.py [hunk 11]
```diff
@@ -285,7 +285,8 @@ def make_exposure(self, obs_id, spectral_index=2.3, for_integral_flux=False):
...
+        center = exposure.center()
```
Comment (JouvinLea) [inline gammapy/scripts/image_pipe.py]: @adonath 
Do you agree now with the test I added that the center() method of the Skymap is ok here?

**Comment by JouvinLea** [comment]
@adonath 
I don't understand why the test don't pass on travis because it passes locally on my machine...

File: gammapy/scripts/image_pipe.py (hunk 11 above)
Comment (adonath) [inline gammapy/scripts/image_pipe.py]: Yes, looks fine to me!

**Comment by adonath** [comment]
@JouvinLea I assume it's related to the subprocess calls, but I can't say for sure. Is there a way to avoid these? E.g. by using `tmpdir.mkdir()` etc.?

@cdeil Any further ideas why the test fails of Travis-CI?

**Comment by joleroi** [comment]
You could avoid using ``subprocess`` by creating the DataStore from ``gammapy-extra`` and copying it to ``tmpdir`` with the ``copy_obs`` command

```
ds = DataStore.from_dir($GAMMAPY_EXTRA/etc.)
ds.copy_obs(ds.obs_table, tmpdir)
```

**Comment by JouvinLea** [comment]
@adonath 
I did what @joleroi proposed!
Now we have an error coming from reproject in skymap in travis
```
from reproject import reproject_interp, reproject_exact
```

**Comment by adonath** [comment]
Great, this already looks much better on Travis-CI! I think you have to  add `@requires_dependency('reproject')` to the `test_image_pipe` function, because it uses the `SkyMap.reproject()` method to reproject the exclusion mask. Once this is done it should be fine. Thanks for your work!
